### ISSUE TITLE
Add check for missing call to `super().__init__` in subclass

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#2789]
When a parent class (other than `object`) defines an `__init__` method that is not called by a subclass from its own `__init__` method, an error should probably be generated.

**Comment by erictraut** [comment]
This is related to [this mypy issue](https://github.com/python/mypy/issues/9225).

```python
class Base():
    def __init__(self, x: int) -> None:
        self.x = x
    def whatsx(self) -> int:
        return self.x

class Child(Base):
    def __init__(self, x: int) -> None:
        pass
        # Note: no call to super(), nor defining self.x.

c = Child(3)
print(c.whatsx())  # mypy does not catch undefined c.x
```

**Comment by NeilGirdhar** [comment]
This is a great idea.  In my opinion, this also applies to `__init_subclass__`, `__enter__`, and `__exit__`.

**Comment by erictraut** [comment]
I don't think it's feasible for a static analyzer to catch undefined attributes in the subclass. A missing call to `super().__init__` is a good proxy for this common error. 

I agree that this would need to be controlled by a separate diagnostic check, and it should probably be disabled by default in "basic" type checking mode, but probably enabled in "strict" mode.

**Comment by erictraut** [comment]
I've added a new diagnostic check called "reportMissingSuperCall". It's currently applied to `__init__`, `__init_subclass__`, `__enter__`, and `__exit__` methods. It's off by default in basic type checking mode but on in strict mode. This will be included in the next release.

**Comment by erictraut** [comment]
After applying this to an internal "strictly typed" code base, I'm worried that it may be too disruptive to enable this check by default for strict mode, so I'm going to be a bit conservative and leave it off by default, even for strict mode for now. We can always add it back in to the strict rule set later.

**Comment by erictraut** [comment]
This is included in pyright 1.1.209, which I just published. It will also be included in the next release of pylance.
